java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -97,12 +99,28 @@ public void visitNode(Tree tree) {
     if (componentScanValues != null) {
       componentScanValues.forEach(this::addToScannedPackages);
     } else if (hasAnnotation(classSymbolMetadata, SPRING_BOOT_APP_ANNOTATION)) {
-      packagesScannedBySpring.add(classPackageName);
+      packagesScannedBySpring.addAll(targetedPackages(classPackageName, classSymbolMetadata));
     } else if (hasAnnotation(classSymbolMetadata, SPRING_BEAN_ANNOTATIONS)) {
       addMessageToMap(classPackageName, classTree.simpleName());
     }
   }
 
+  private static List<String> targetedPackages(String classPackageName, SymbolMetadata classSymbolMetadata) {
+    // annotation is necessarily there already
+    return Objects.requireNonNull(classSymbolMetadata.valuesForAnnotation(SPRING_BOOT_APP_ANNOTATION))
+      .stream()
+      .filter(v -> "scanBasePackages".equals(v.name()))
+      .map(SymbolMetadata.AnnotationValue::value)
+      .findFirst()
+      // list of packages to scan
+      .filter(Object[].class::isInstance)
+      .map(Object[].class::cast)
+      .map(Arrays::asList)
+      .map(list -> list.stream().map(String.class::cast).collect(Collectors.toList()))
+      // package in which the annotated class is defined
+      .orElse(Collections.singletonList(classPackageName));
```

Comment:
This matches my understanding of the annotation and seems correct. We could be a bit clearer in the comment, to say that this is the default behavior of the annotation, for example:


Using this annotation without arguments tells Spring to scan the current package and all of its sub-packages.